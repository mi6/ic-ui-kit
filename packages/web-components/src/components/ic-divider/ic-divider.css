hr {
  padding: 0;
  margin: 0;
  border: none;
  height: var(--ic-space-1px);
  background-color: var(--ic-keyline-darken-rgb);
}

hr.light {
  background-color: var(--ic-keyline-lighten-rgb);
}
