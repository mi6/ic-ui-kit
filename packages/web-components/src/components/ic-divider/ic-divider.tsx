import {
  h,
  Component,
  Host,
  Listen,
  State,
  Prop,
  Element,
} from "@stencil/core";

import {
  getBrandForegroundAppearance,
  isPropDefined,
  slotHasContent,
} from "../../utils/helpers";

import {
  IcBrand,
  IcOrientation,
  IcBrandForeground,
  IcThemeMode,
} from "../../utils/types";

import {
  IcDividerLabelPlacement,
  IcDividerStyles,
  IcDividerWeights,
} from "./ic-divider.types";

@Component({
  tag: "ic-divider",
  styleUrl: "ic-divider.css",
  scoped: true,
})
export class Divider {
  private parentElIsSideNav: boolean = false;

  @Element() el: HTMLIcDividerElement;

  @State() foregroundColor: IcBrandForeground = getBrandForegroundAppearance();

  /**
   * The line style of the divider.
   */
  @Prop() borderStyle?: IcDividerStyles = "solid";

  /**
   * The label for the divider. The label placement will need to be set for the label to be displayed correctly.
   */
  @Prop() label?: string;

  /**
   * The position the label is placed on the divider. `Left` and `right` placement is only applicable when orientation is set to `horizontal`. `Top` and `bottom` placement is only applicable when orientation is set to `vertical`. `Center` placement is applicable for both orientations.
   */
  @Prop() labelPlacement?: IcDividerLabelPlacement = "center";

  /** If `true`, the divider will be displayed in a grey colour. */
  @Prop() monochrome?: boolean = false;

  /**
   * The orientation of the divider.
   */
  @Prop() orientation?: IcOrientation = "horizontal";

  /**
   * Sets the theme color to the dark or light theme color. "inherit" will set the color based on the system settings or ic-theme component.
   */
  @Prop({ mutable: true }) theme?: IcThemeMode | IcBrandForeground = "inherit";

  /**
   * The thickness of the divider.
   */
  @Prop() weight?: IcDividerWeights = "thin";

  @Listen("brandChange", { target: "document" })
  brandChangeHandler(ev: CustomEvent<IcBrand>): void {
    this.foregroundColor = ev.detail.mode;
  }

  private updateMonochromeState(): void {
    if (this.parentElIsSideNav) {
      this.el.classList.add("ic-side-navigation-keyline");
      if (this.foregroundColor === "light") {
        this.theme = "dark";
      } else {
        this.theme = "light";
      }
    }
  }

  componentWillRender(): void {
    const parentEl = this.el.parentElement;
    if (parentEl) {
      const isBottomSideNav = parentEl.classList.contains("bottom-side-nav");
      const isBottomWrapper =
        parentEl.parentElement?.classList.contains("bottom-wrapper");

      if (
        parentEl.tagName === "IC-SIDE-NAVIGATION" ||
        (isBottomSideNav && isBottomWrapper)
      ) {
        this.parentElIsSideNav = true;
      }
    }
    this.updateMonochromeState();
  }

  render() {
    const {
      borderStyle,
      label,
      labelPlacement,
      monochrome,
      orientation,
      weight,
    } = this;

    const { theme } = this;

    const invalidPropCombination = (
      placement: IcDividerLabelPlacement
    ): void => {
      console.error(
        `${
          placement.charAt(0).toUpperCase() + placement.slice(1)
        } label placement is not applicable for a ${orientation} ic-divider (web-components) / IcDivider (react)`
      );
    };

    if (
      ((labelPlacement === "left" || labelPlacement === "right") &&
        orientation === "vertical") ||
      ((labelPlacement === "top" || labelPlacement === "bottom") &&
        orientation === "horizontal")
    ) {
      invalidPropCombination(labelPlacement);
    }

    const getTypographyVariant = (weight: string) => {
      switch (weight) {
        case "very-thick":
          return "h4";
        case "thick":
          return "subtitle-large";
        default:
          return "label";
      }
    };

    const renderLabel = () => {
      if (slotHasContent(this.el, "label")) {
        return (
          <div class="ic-divider-label-container">
            <slot name="label"></slot>
          </div>
        );
      } else if (isPropDefined(label)) {
        return (
          <ic-typography
            class="ic-divider-label"
            variant={getTypographyVariant(weight!)}
          >
            <p>{label}</p>
          </ic-typography>
        );
      }
      return null;
    };

    return (
      <Host
        class={{
          [`ic-theme-${theme}`]: theme !== "inherit",
          [`ic-divider-monochrome`]: !!monochrome,
          [`ic-divider-${orientation}`]: true,
          [`ic-divider-${weight}`]: true,
          [`ic-divider-${borderStyle}`]: true,
        }}
        {...((orientation === "vertical" ||
          (!!renderLabel() && !!isPropDefined(labelPlacement))) && {
          "aria-orientation": orientation,
          role: "separator",
        })}
      >
        {orientation === "horizontal" &&
          (!renderLabel() || !isPropDefined(labelPlacement)) &&
          (this.parentElIsSideNav ? <li role="separator" /> : <hr />)}
        {!!isPropDefined(labelPlacement) && !!renderLabel() && renderLabel()}
        {orientation === "vertical" &&
          (!renderLabel() || !isPropDefined(labelPlacement)) && (
            <div class="vertical-divider"></div>
          )}
      </Host>
    );
  }
}
