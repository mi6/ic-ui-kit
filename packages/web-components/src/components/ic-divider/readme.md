# ic-divider



<!-- Auto Generated Below -->


## Properties

| Property         | Attribute         | Description                                                                                                                                                                                                                                                                              | Type                                                 | Default        |
| ---------------- | ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------- | -------------- |
| `borderStyle`    | `border-style`    | The line style of the divider.                                                                                                                                                                                                                                                           | `"dashed" \| "solid"`                                | `"solid"`      |
| `label`          | `label`           | The label for the divider. The label placement will need to be set for the label to be displayed correctly.                                                                                                                                                                              | `string`                                             | `undefined`    |
| `labelPlacement` | `label-placement` | The position the label is placed on the divider. `Left` and `right` placement is only applicable when orientation is set to `horizontal`. `Top` and `bottom` placement is only applicable when orientation is set to `vertical`. `Center` placement is applicable for both orientations. | `"bottom" \| "center" \| "left" \| "right" \| "top"` | `"center"`     |
| `monochrome`     | `monochrome`      | If `true`, the divider will be displayed in a grey colour.                                                                                                                                                                                                                               | `boolean`                                            | `false`        |
| `orientation`    | `orientation`     | The orientation of the divider.                                                                                                                                                                                                                                                          | `"horizontal" \| "vertical"`                         | `"horizontal"` |
| `theme`          | `theme`           | Sets the theme color to the dark or light theme color. "inherit" will set the color based on the system settings or ic-theme component.                                                                                                                                                  | `"dark" \| "default" \| "inherit" \| "light"`        | `"inherit"`    |
| `weight`         | `weight`          | The thickness of the divider.                                                                                                                                                                                                                                                            | `"medium" \| "thick" \| "thin" \| "very-thick"`      | `"thin"`       |


## CSS Custom Properties

| Name                                 | Description                                                                                                                                                                                |
| ------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `--ic-divider-background`            |                                                                                                                                                                                            |
| `--ic-divider-background-monochrome` |                                                                                                                                                                                            |
| `--ic-divider-horizontal-width`      | The width of a horizontal divider. The default value is `inherit`.                                                                                                                         |
| `--ic-divider-label`                 |                                                                                                                                                                                            |
| `--ic-divider-label-monochrome`      |                                                                                                                                                                                            |
| `--ic-divider-label-width`           | The max-width of the divider label. When orientation is set to `horizontal`, the default value is `max(20rem, 60%)`. When orientation is set to `vertical`, the default value is `20 rem`. |
| `--ic-divider-vertical-height`       | The height of a vertical divider. The default value is `inherit`.                                                                                                                          |


## Dependencies

### Used by

 - [ic-side-navigation](../ic-side-navigation)

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-divider --> ic-typography
  ic-side-navigation --> ic-divider
  style ic-divider fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


