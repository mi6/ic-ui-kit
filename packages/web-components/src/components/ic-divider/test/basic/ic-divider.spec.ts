import { newSpecPage } from "@stencil/core/testing";
import { Divider } from "../../ic-divider";

describe("ic-divider", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-divider></ic-divider>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("correctly sets foregroundColor on theme change", async () => {
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-divider></ic-divider>`,
    });

    await page.rootInstance.themeChangeHandler({
      detail: { mode: "light" },
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();

    expect(page.rootInstance.foregroundColor).toEqual("light");
  });
});
