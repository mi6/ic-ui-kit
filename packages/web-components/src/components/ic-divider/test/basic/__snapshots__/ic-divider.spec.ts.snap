// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-divider correctly sets foregroundColor on brand change: brand-change-dark 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin ic-side-navigation-keyline ic-theme-light" slot="primary-navigation">
  <!---->
  <li role="separator"></li>
</ic-divider>
`;

exports[`ic-divider correctly sets foregroundColor on brand change: brand-change-light 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin ic-side-navigation-keyline ic-theme-dark" slot="primary-navigation">
  <!---->
  <li role="separator"></li>
</ic-divider>
`;

exports[`ic-divider should render 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin">
  <!---->
  <hr>
</ic-divider>
`;

exports[`ic-divider should render a label with the h4 typography variant when weight is set to very-thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-very-thick" label="Test label" role="separator" weight="very-thick">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-h4">
    <template shadowrootmode="open">
      <slot></slot>
    </template>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render a label with the label typography variant when weight is not set to thick or very-thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-medium ic-divider-solid" label="Test label" role="separator" weight="medium">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-label">
    <template shadowrootmode="open">
      <slot></slot>
    </template>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render a label with the subtitle-large typography variant when weight is set to thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-thick" label="Test label" role="separator" weight="thick">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-subtitle-large">
    <template shadowrootmode="open">
      <slot></slot>
    </template>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render with a slotted label 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-thin" label-placement="left" role="separator">
  <!---->
  <div class="ic-divider-label-container">
    <ic-typography slot="label" variant="label">
      <p style="width: max-content;">
        Slotted label left
      </p>
    </ic-typography>
  </div>
</ic-divider>
`;
